test/distributed/fsdp/test_checkpoint_wrapper.py

```diff
@@ -189,5 +189,47 @@ def test_fqn(self):
         for fqn, _ in lin.named_parameters():
             self.assertTrue(fqn in state_dict, msg=f"{fqn} not in state_dict.")
 
+    def test_checkpoint_wrapper_cpu_offload(self):
+        model = nn.Sequential(
+            nn.Linear(10, 10),
+            nn.Linear(10, 10),
+            nn.Linear(10, 10),
+        ).cuda()
+
+        # Patch saved_tensor_hooks to make the unpack keep the tensor on CPU for
+        # testing, otherwise the tensor access during the DFS will cause orig
+        # unpack to run, transferring the tensor back to GPU.
+        def patched_init(saved_tensor_hook_obj, pack_hook, _):
+            saved_tensor_hook_obj.pack_hook = pack_hook
+
+            def testing_cpu_offload_unpack_hook(packed):
+                _, tensor = packed
+                return tensor
+
+            saved_tensor_hook_obj.unpack_hook = testing_cpu_offload_unpack_hook
+
+        torch.autograd.graph.saved_tensors_hooks.__init__ = patched_init
+
+        model = checkpoint_wrapper(model, offload_to_cpu=True)
+
+        inp = torch.randn(3, 10, device='cuda')
+        loss = model(inp).sum()
+
+        # All autograd saved tensors should be offloaded to CPU.
+        def dfs(grad_fn):
+            for e in dir(grad_fn):
+                if not e.startswith('_saved_'):
+                    continue
+
+                saved = getattr(grad_fn, e)
+                if isinstance(saved, torch.Tensor):
+                    self.assertEqual(torch.device("cpu"), saved.device)
+
+            if hasattr(grad_fn, 'next_functions'):
+                for k, _ in grad_fn.next_functions:
```

Comment:
nit: Does `k` stand for something? Maybe we can use a more descriptive name here since the 1st element of each tuple in `next_functions` is always a backward function (I think)?